### PR TITLE
server: wrap etcd operations

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Ref #5836.

### What is changed and how does it work?
This PR adds the basic watch function for finding the key changes which might be used in the client. And also support put and get functions.
Here is the proto definition: https://github.com/pingcap/kvproto/pull/1061

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -1905,6 +1949,104 @@ func (c *client) WatchGlobalConfig(ctx context.Context, configPath string, revis
...
+	}
+	start := time.Now()
+	defer func() { cmdDurationPut.Observe(time.Since(start).Seconds()) }()
+
+	ctx, cancel := context.WithTimeout(ctx, c.option.timeout)
+	req := &pdpb.PutRequest{
+		Key:    key,
+		Value:  value,
+		Lease:  options.lease,
+		PrevKv: options.prevKv,
```
Comment (binshi-bing) [inline client/client.go:1970]: Who sets the prevKv to options?

**Review by binshi-bing — COMMENTED** [review]
left some comments.

File: server/grpc_service.go [hunk 2]
```diff
@@ -2007,6 +2007,111 @@ func (s *GrpcServer) WatchGlobalConfig(req *pdpb.WatchGlobalConfigRequest, serve
...
 }
 
+// Watch watches the key with a given prefix and revision.
+func (s *GrpcServer) Watch(req *pdpb.WatchRequest, server pdpb.PD_WatchServer) error {
+	ctx, cancel := context.WithCancel(s.Context())
+	defer cancel()
+	key := string(req.GetKey())
+	endKey := string(req.GetRangeEnd())
+	startRevision := req.GetStartRevision()
+	watchChan := s.client.Watch(ctx, key, clientv3.WithPrefix(), clientv3.WithRange(endKey), clientv3.WithRev(startRevision), clientv3.WithPrevKV())
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/grpc_service.go:2017]: This PR is going to replace the global config related RPC. And pitr will use it later. BTW, the microservice will also use it.

File: server/grpc_service.go [hunk 3]
```diff
@@ -2007,6 +2007,111 @@ func (s *GrpcServer) WatchGlobalConfig(req *pdpb.WatchGlobalConfigRequest, serve
...
+	if endKey := req.GetRangeEnd(); endKey != nil {
+		options = append(options, clientv3.WithRange(string(endKey)))
+	}
+	if rev := req.GetRevision(); rev != 0 {
+		options = append(options, clientv3.WithRev(rev))
+	}
+	if limit := req.GetLimit(); limit != 0 {
+		options = append(options, clientv3.WithLimit(limit))
+	}
+	res, err := s.client.Get(ctx, key, options...)
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/grpc_service.go:2072]: It's just a simple wrapper for etcd API.

File: client/client.go (hunk 1 above)
Comment (rleungx) [reply client/client.go:1970]: etcd supports it so I decide to keep it

File: client/client.go [hunk 4]
```diff
@@ -1905,6 +1949,104 @@ func (c *client) WatchGlobalConfig(ctx context.Context, configPath string, revis
...
+	resp, err := c.getClient().Get(ctx, req)
+	cancel()
+
+	if err = c.respForErr(cmdFailedDurationGet, start, err, resp.GetHeader()); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (c *client) Watch(ctx context.Context, key []byte, opts ...OpOption) (chan []*pdpb.Event, error) {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply client/client.go:2012]: Yes, I have taken it into consideration. And we need a discussion about whether we need to put them into a new pb.

File: server/grpc_service.go [hunk 5]
```diff
@@ -2007,6 +2007,111 @@ func (s *GrpcServer) WatchGlobalConfig(req *pdpb.WatchGlobalConfigRequest, serve
 	}
 }
 
+// Watch watches the key with a given prefix and revision.
+func (s *GrpcServer) Watch(req *pdpb.WatchRequest, server pdpb.PD_WatchServer) error {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/grpc_service.go:2011]: same with https://github.com/tikv/pd/pull/6025#discussion_r1115182764

File: client/client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rleungx) [reply client/client.go:2012]: After an offline discussion, we may want to add some custom fields in the request, it's more flexible to put them into pdpb.

File: client/client.go [hunk 6]
```diff
@@ -1905,6 +1949,104 @@ func (c *client) WatchGlobalConfig(ctx context.Context, configPath string, revis
 	return globalConfigWatcherCh, err
 }
 
+func (c *client) Put(ctx context.Context, key, value []byte, opts ...OpOption) (*pdpb.PutResponse, error) {
+	options := &Op{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if span := opentracing.SpanFromContext(ctx); span != nil {
```
Comment (bufferflies) [inline client/client.go:1958]: how about exract the trace code as one funtion?

Comment (rleungx) [reply client/client.go]: Will do it in another PR

**Comment by rleungx** [comment]
@YuJuncen PTAL

File: server/meta_storage_service.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/pingcap/kvproto/pkg/meta_storagepb"
+	"go.etcd.io/etcd/clientv3"
+)
+
+// MetaStorageServer wraps GrpcServer to provide meta storage service.
+type MetaStorageServer struct {
```
Comment (nolouch) [inline server/meta_storage_service.go:26]: What about moving the `metastoregeserver` to `msc`?

**Review by nolouch — COMMENTED** [review]
I think @glorv may have some concerns, ptal @glorv

**Review by glorv — COMMENTED** [review]
LGTM

File: server/meta_storage_service.go (hunk 7 above)
Comment (rleungx) [reply server/meta_storage_service.go]: done

File: pkg/mcs/meta_storage/server/grpc_service.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+	if err := s.checkServing(); err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(s.ctx)
+	defer cancel()
+	key := string(req.GetKey())
+	endKey := string(req.GetRangeEnd())
+	startRevision := req.GetStartRevision()
+	cli := s.manager.GetClient()
+	watchChan := cli.Watch(ctx, key, clientv3.WithPrefix(), clientv3.WithRange(endKey), clientv3.WithRev(startRevision), clientv3.WithPrevKV())
```
Comment (nolouch) [inline pkg/mcs/meta_storage/server/grpc_service.go:95]: Do we need to let the client can pass the options? such as some clients may no need `WithPrevKV`.

Comment (rleungx) [reply pkg/mcs/meta_storage/server/grpc_service.go]: Done, and remove with prefix since it should be handled in the client side.

**Review by nolouch — APPROVED** [review]
lgtm

File: pkg/mcs/discovery/register.go [hunk 9]
```diff
@@ -45,7 +45,7 @@ func NewServiceRegister(ctx context.Context, cli *clientv3.Client, serviceName,
 		key:    serviceKey,
 		value:  serializedValue,
 		ttl:    ttl,
-	}
+	}, nil
```
Comment (bufferflies) [inline pkg/mcs/discovery/register.go:48]: when it return err?

File: pkg/mcs/meta_storage/server/manager.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+	bs "github.com/tikv/pd/pkg/basicserver"
+	"github.com/tikv/pd/pkg/storage/endpoint"
+	"github.com/tikv/pd/pkg/storage/kv"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+)
+
+// Manager is the manager of resource group.
+type Manager struct {
+	srv       bs.Server
```
Comment (bufferflies) [inline pkg/mcs/meta_storage/server/manager.go:28]: where does it use?

Comment (rleungx) [reply pkg/mcs/meta_storage/server/manager.go:28]: IsServing will use it.

File: client/meta_storage_client.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+	})
+	if err != nil {
+		close(eventCh)
+		return nil, err
+	}
+	go func() {
+		defer func() {
+			close(eventCh)
+			if r := recover(); r != nil {
+				log.Error("[pd] panic in client `Watch`", zap.Any("error", r))
```
Comment (bufferflies) [inline client/meta_storage_client.go:194]: plz use zap.Error

File: client/meta_storage_client.go [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+			case <-ctx.Done():
+				return
+			case eventCh <- resp.GetEvents():
+			}
+		}
+	}()
+	return eventCh, err
+}
+
+func (c *client) respForMetaStorageErr(observer prometheus.Observer, start time.Time, err error, header *meta_storagepb.ResponseHeader) error {
```
Comment (bufferflies) [inline client/meta_storage_client.go:213]: what's the difference with `client.respForErr`

File: client/meta_storage_client.go [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+		if end[i] < 0xff {
+			end[i]++
+			end = end[:i+1]
+			return end
+		}
+	}
+	return []byte{0}
+}
+
+func (c *client) Put(ctx context.Context, key, value []byte, opts ...OpOption) (*meta_storagepb.PutResponse, error) {
```
Comment (bufferflies) [inline client/meta_storage_client.go:107]: I think the implement's code should move to client.go not belongs here.

**Review by bufferflies — COMMENTED** [review]
The rest LGTM

File: client/meta_storage_client.go (hunk 12 above)
Comment (rleungx) [reply client/meta_storage_client.go:213]: Almost the same, only the header is defined in a different proto, maybe we can use generic to do some refactoring later.

File: client/meta_storage_client.go (hunk 13 above)
Comment (rleungx) [reply client/meta_storage_client.go:107]: It's the same as the keyspace client and resource manager client. I think it's more clear.

File: tests/client/client_test.go [hunk 14]
```diff
@@ -1422,3 +1424,71 @@ func (suite *clientTestSuite) TestScatterRegion() {
...
+	getResp, err := client.Get(context.Background(), key)
+	re.NoError(err)
+	re.Equal([]byte("1"), getResp.GetKvs()[0].Value)
+	re.NotEqual(0, getResp.GetHeader().GetRevision())
+	putResp, err = client.Put(context.Background(), key, []byte("2"), pd.WithPrevKV())
+	re.NoError(err)
+	re.Equal([]byte("1"), putResp.GetPrevKv().Value)
+	getResp, err = client.Get(context.Background(), key)
+	re.NoError(err)
+	re.Equal([]byte("2"), getResp.GetKvs()[0].Value)
```
Comment (bufferflies) [inline tests/client/client_test.go:1493]: how about add delete key test?

Comment (rleungx) [reply tests/client/client_test.go:1494]: Added

**Review by bufferflies — APPROVED** [review]

**Comment by lhy1024** [comment]
TestWatch is timed out after 10m

https://github.com/tikv/pd/actions/runs/4292184591/jobs/7478308565

**Pull request merged**